### DISCUSSION TITLE
Pico W http server slow response

### DISCUSSION THREAD

**Discussion by gwynge** [micropython/micropython#9922] [General]
I'm using a Pico W as a monitor with an http server running via asyncio.start_server.

Every so often http request goes from taking perhaps 200mS to taking 15+ seconds. It does eventually respond. I'm struggling to reproduce when connected to Thonny console which is a bit annoying.

Anybody seen this sort of behaviour and have any wisdom?

**Comment by n1kdo** [comment]
do you ever call gc.collect() in your program to force garbage collection to occur -- when you want it to, rather than when the system decides to do it?

**Reply by gwynge** [reply]
I looked at memory usage, it fills up and collects and couldn't make anything happen in that respect so suspect it's not related (to this particular behaviour anyway)

**Comment by fivdi** [comment]
Maybe the power-saving mode on the wireless chip which is activated by default is causing the server to be less responsive. For further information see section "3.6.3. Power-saving mode" of [Connecting to the Internet with the raspberry Pi Pico W](https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf).

**Reply by gwynge** [reply]
I'm sure I'd read the documentation before but not seen any mention of power save mode, Wonder if I found an old document? It also defines the response code of 2 when connecting which I couldn't find a definition of anywhere else!
I shall try disabling powersave and see if that works. I wondered if there was such a feature. Although I note that "idle" is not defined. In other words how long before it goes into power save mode?

**Comment by n1kdo** [comment]
I can serve static html pages from Pico-W in < .002 sec., so something is not right.

Here is the no-powersave code that I use to start the network.

```
        print('Connecting to WLAN...')
        wlan = network.WLAN(network.STA_IF)
        wlan.config(pm=0xa11140)  # disable power save, this is a server.
        # wlan.config(hostname='not supported on pico-w')
        wlan.active(True)
        wlan.connect(ssid, secret)
```

**Reply by gwynge** [reply]
Thanks, I shall try this.

**Comment by DavesCodeMusings** [comment]
I'm assuming you're coding your own http server here and not using a library. If so, use `netstat` on the client side to check for tcp connections in a TIME_WAIT or FIN_WAIT state. I had a similar problem where requests were not ended properly and the http server was running out of connections. That was what caused the odd, intermittent hanging for me. I think the default number of connections with `start_server` is 5, so if you see 5 or more waiting in the `netstat` output, you know what the problem is. 

Here's a link to the details of my original problem and its solution. https://github.com/micropython/micropython/discussions/9695
